lib/include/prjxray/xilinx/frames.h

```diff
@@ -29,15 +33,88 @@ class Frames {
 	    const absl::optional<typename ArchType::Part>& part);
 
 	// Returns the map with frame addresses and corresponding data
-	Frames2Data& getFrames();
+	Frames2Data& getFrames() { return frames_data_; }
 
        private:
 	Frames2Data frames_data_;
-
-	// Updates the ECC information in the frame.
-	void updateECC(FrameData& data);
 };
 
+template <typename ArchType>
+int Frames<ArchType>::readFrames(const std::string& frm_file_str) {
+	assert(!frm_file_str.empty());
+
+	std::ifstream frm_file(frm_file_str);
+	if (!frm_file) {
+		std::cerr << "Unable to open frm file: " << frm_file_str
+		          << std::endl;
+		return 1;
+	}
+	std::string frm_line;
+
+	while (std::getline(frm_file, frm_line)) {
+		if (frm_line[0] == '#')
+			continue;
+
+		std::pair<std::string, std::string> frame_delta =
+		    absl::StrSplit(frm_line, ' ');
+
+		uint32_t frame_address =
+		    std::stoul(frame_delta.first, nullptr, 16);
+
+		std::vector<std::string> frame_data_strings =
+		    absl::StrSplit(frame_delta.second, ',');
+
+		if (frame_data_strings.size() != ArchType::words_per_frame) {
+			std::cerr << "Frame " << std::hex << frame_address
+			          << ": found " << std::dec
+			          << frame_data_strings.size()
+			          << " words instead of "
+			          << ArchType::words_per_frame << std::endl;
+			continue;
+		}
+
+		FrameData frame_data(frame_data_strings.size(), 0);
+		std::transform(frame_data_strings.begin(),
+		               frame_data_strings.end(), frame_data.begin(),
+		               [](const std::string& val) -> uint32_t {
+			               return std::stoul(val, nullptr, 16);
+		               });
+
+		// Spartan6 doesn't have ECC
+		if (std::is_same<ArchType, Series7>::value) {
```

Comment:
Just have `Spartan6` specialized updateECC to a `nop`.

Reply:
In this PR I removed the type checking and added a generic updateECC template method.
I will add the specialization in the Spartan6 related PR - see [commit](https://github.com/SymbiFlow/prjxray/pull/1088/commits/083363f5806e5b405cd61d8b9869b37986f32a87)